pkg/kubelet/kubelet.go

```diff
@@ -1258,12 +1264,40 @@ func makePortMappings(container *api.Container) (ports []kubecontainer.PortMappi
 	return
 }
 
+func generatePodHostNameAndDomain(pod *api.Pod, clusterDomain string) (string, string) {
+	// TODO(vmarmol): Handle better.
+	// Cap hostname at 63 chars (specification is 64bytes which is 63 chars and the null terminating char).
+	const hostnameMaxLen = 63
+	podAnnotations := pod.Annotations
+	if podAnnotations == nil {
+		podAnnotations = make(map[string]string)
+	}
+	hostname := podAnnotations[utilpod.PodHostnameAnnotation]
+	if !utilvalidation.IsDNS1123Label(hostname) {
```

Comment:
actually, if we are calling this "beta" we should do validation in the API validation pass, which will reject pods with invalid hostnames or domain.